File: web/backend/api/survival.py
# SPDX-License-Identifier: AGPL-3.0
from typing import Any

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

router = APIRouter(prefix="/api/survival", tags=["survival"])


class SurvivalEstimateRequest(BaseModel):
    char_data: dict[str, Any]
    weapon_data: dict[str, Any]
    char_level: int = 90
    weapon_level: int = 90
    trust_level: int = 0
    enemy_tier: str = "普通"
    imbalance_efficiency_bonus: float = 0.0
    enemy_max_hp: float | None = None
    enemy_id: str = ""
    base_heal_flat: float = 201.6
    stat_per_point: float = 0.47
    heal_efficiency: float = 0.20
    independent_heal_bonus: float = 0.30
    imbalance_gain_base: float = 10.0
    hot_resistance_percent: float = 0.0
    sp_start: float = 0.0
    sp_seconds: float = Field(5.0, ge=0.0)
    ult_start: float = 0.0
    life_steal_rate: float = Field(0.10, ge=0.0, le=1.0)


@router.post("/estimate")
def survival_estimate(req: SurvivalEstimateRequest) -> dict[str, Any]:
    from games.endfield.calc.survival.estimate import build_survival_estimate

    try:
        return build_survival_estimate(
            char_data=req.char_data,
            weapon_data=req.weapon_data,
            char_level=req.char_level,
            weapon_level=req.weapon_level,
            trust_level=req.trust_level,
            enemy_tier=req.enemy_tier,
            imbalance_efficiency_bonus=req.imbalance_efficiency_bonus,
            enemy_max_hp=req.enemy_max_hp,
            enemy_id=req.enemy_id,
            base_heal_flat=req.base_heal_flat,
            stat_per_point=req.stat_per_point,
            heal_efficiency=req.heal_efficiency,
            independent_heal_bonus=req.independent_heal_bonus,
            imbalance_gain_base=req.imbalance_gain_base,
            hot_resistance_percent=req.hot_resistance_percent,
            sp_start=req.sp_start,
            sp_seconds=req.sp_seconds,
            ult_start=req.ult_start,
            life_steal_rate=req.life_steal_rate,
        )
    except Exception as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc

File: web/frontend/src/api/survival.ts
import i18n from "../i18n/config";

const BASE = "/api/survival";

export interface SurvivalEstimateRequest {
  char_data: Record<string, unknown>;
  weapon_data: Record<string, unknown>;
  char_level?: number;
  weapon_level?: number;
  trust_level?: number;
  enemy_tier?: string;
  imbalance_efficiency_bonus?: number;
  enemy_max_hp?: number | null;
  enemy_id?: string;
  base_heal_flat?: number;
  stat_per_point?: number;
  heal_efficiency?: number;
  independent_heal_bonus?: number;
  imbalance_gain_base?: number;
  hot_resistance_percent?: number;
  sp_start?: number;
  sp_seconds?: number;
  ult_start?: number;
  life_steal_rate?: number;
}

export interface SurvivalEstimateResult {
  execute_damage: number;
  execute_multiplier: number;
  execute_sp_restore: number;
  imbalance_cap: number;
  imbalance_duration_sec: number;
  imbalance_nodes_1: number[];
  imbalance_nodes_2: number[];
  imbalance_gain_effective: number;
  imbalance_gain_percent: number;
  fast_break_multiplier: number;
  burn_tick_per_sec: number;
  enemy_max_hp: number;
  sp_after_regen: number;
  sp_regen_per_sec: number;
  ultimate_charge_after: number;
  ultimate_charge_per_100_sp: number;
  dodge_sp_gain: number;
  life_steal_heal: number;
  healing_amount: number;
  character_max_hp: number;
}

export async function fetchSurvivalEstimate(
  params: SurvivalEstimateRequest,
): Promise<SurvivalEstimateResult> {
  const r = await fetch(`${BASE}/estimate`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(params),
  });
  if (!r.ok) {
    const text = await r.text();
    throw new Error(`${i18n.t("api.survivalEstimateFailed")}: ${text}`);
  }
  return r.json();
}
